scripts/cnode-helper-scripts/cncli.sh

```diff
@@ -794,10 +801,173 @@ cncliPTsendslots() {
   done
 }
 
+#################################
+# epochdata table load process  #
+#################################
+
+getCurrNextEpoch() {
+  getNodeMetrics
+  curr_epoch=${epochnum}
+  next_epoch=$((curr_epoch+1))
+}
+
+runCurrentEpoch() {
+  getKoiosData
+  echo "Processing current epoch: ${1}"
+  stake_param_curr="--active-stake ${active_stake_set} --pool-stake ${pool_stake_set}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${1}" ${stake_param_curr} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+}
+
+runNextEpoch() {
+  getKoiosData
+  echo "Processing next epoch: ${1}"
+  stake_param_next="--active-stake ${active_stake_mark} --pool-stake ${pool_stake_mark}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${1}" ${stake_param_next} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+}
+
+runPreviousEpochs() {
+  [[ -z ${KOIOS_API} ]] && return 1
+  if ! pool_hist=$(curl -sSL -f "${KOIOS_API}/pool_history?_pool_bech32=${POOL_ID_BECH32}&_epoch_no=${1}" 2>&1); then
+    echo "ERROR: Koios pool_stake_snapshot history query failed."
+    return 1
+  fi
+
+  if ! epoch_hist=$(curl -sSL -f "${KOIOS_API}/epoch_info?_epoch_no=${1}" 2>&1); then
+    echo "ERROR: Koios epoch_stake_snapshot history query failed."
+    return 1
+  fi
+
+  pool_stake_hist=$(jq -r '.[].active_stake' <<< "${pool_hist}")
+  active_stake_hist=$(jq -r '.[].active_stake' <<< "${epoch_hist}")
+
+  echo "Processing previous epoch: ${1}"
+  stake_param_prev="--active-stake ${active_stake_hist} --pool-stake ${pool_stake_hist}"
+
+  ${CNCLI} leaderlog ${cncliParams} --consensus "${consensus}" --epoch="${1}" ${stake_param_prev} |
+  jq -r '[.epoch, .epochNonce, .poolId, .sigma, .d, .epochSlotsIdeal, .maxPerformance, .activeStake, .totalActiveStake] | @csv' |
+  sed 's/"//g' >> "$tmpcsv"
+
+  return 0
+}
+
+processAllEpochs() {
+  getCurrNextEpoch
+  IFS=' ' read -r -a epochs_array <<< "$EPOCHS"
+
+  for epoch in "${epochs_array[@]}"; do
+    if ! getConsensus "${epoch}"; then echo "ERROR: Failed to fetch protocol parameters for epoch ${epoch}."; return 1; fi
+    if [[ "$epoch" == "$curr_epoch" ]]; then
+      runCurrentEpoch ${epoch}
+    elif [[ "$epoch" == "$next_epoch" ]]; then
+      runNextEpoch ${epoch}
+    else
+      runPreviousEpochs ${epoch}
```

Comment:
For previous epoch, ie current - 1, a function similar to runCurrentEpoch should be added that use `go` values instead of set from getKoiosData. Koios doesn't have data for previous epoch in pool_history endpoint. This might be a bug, looking into it. Nonetheless, this change makes sense. runPreviousEpochs can be used for all older epochs. 

Needs to be added in processSingleEpoch as well.

Reply:
Thank you, I'll review it during the day

Reply:
Hello @Scitz0,

about go field I noticed that it was removed after new Koios release. I just queried pool_history and it returns data for previous epochs.

I'll wait for you input on this point. 

Meanwhile I push the changes on runNextEpoch (added check on stability window).

Cheers,
Manuel

Reply:
pool_stake_snapshot endpoint is affected by the same bug as pool_history. That pool_stats table in dbsync is a bit messed up in current release. 

pool_stake_snapshot should return go snapshot still, if there are data to return. I will think about this for a moment before we finally merge PR.

Reply:
I will just go ahead and merge this as is now. We can possibly switch to go snapshot later.